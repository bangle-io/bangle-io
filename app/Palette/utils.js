import React from 'react';

export function addBoldToTitle(string, query) {
  if (!query) {
    return string;
  }

  let newString = string.split(query);

  return newString.flatMap((r, i) => {
    if (i === 0) {
      return r;
    }
    return [<b key={i}>{query}</b>, r];
  });
}
