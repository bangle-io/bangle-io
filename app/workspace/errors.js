class BaseError extends Error {
  /**
   *
   * @param {*} message
   * @param {*} code - error code
   * @param {*} displayMessage - one that will be shown to the user, generally a non fatal error
   * @param {*} srcError - if error encapsulates another error
   */
  constructor(message, code, displayMessage, srcError) {
    // 'Error' breaks prototype chain here
    super(message);
    // restore prototype chain
    const actualProto = new.target.prototype;
    if (Object.setPrototypeOf) {
      Object.setPrototypeOf(this, actualProto);
    } else {
      this.__proto__ = actualProto;
    }

    if (srcError) {
      console.log('the error occurred while handling this error');
      console.error(srcError);
      this.srcError = srcError;
    }

    if (code) {
      this.code = code;
    }

    this.displayMessage = displayMessage;
    this.name = this.constructor.name;
  }
}

export class FSError extends BaseError {}
export class WorkspaceError extends BaseError {}
