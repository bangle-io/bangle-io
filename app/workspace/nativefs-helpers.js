const DEFAULT_DIR_IGNORE_LIST = ['node_modules', '.git'];
const LOG = true;
let log = LOG ? console.log.bind(console, 'nativefs-helpers') : () => {};

/**
 *
 * @param {string[]} filePath [...parentDirNames, fileName]
 * @param {DirHandle} rootDirHandle
 */
export class NativeFileOps {
  constructor({
    allowedFile,
    allowedDir = (entry) => !DEFAULT_DIR_IGNORE_LIST.includes(entry.name),
  } = {}) {
    this._allowedFile = allowedFile;
    this._allowedDir = allowedDir;
    this._traverseCache = new WeakMap();
    this._getFileHandle = getFileHandle({ allowedDir, allowedFile });
  }

  _calculateFilesCache = async (rootDirHandle) => {
    const listOfFiles = await recurseDirHandle(rootDirHandle, {
      allowedFile: this._allowedFile,
      allowedDir: this._allowedDir,
    });

    const availableFiles = new Map(
      listOfFiles.map((f) => {
        return [f.map((r) => r.name).join('/'), f];
      }),
    );

    this._traverseCache.set(rootDirHandle, availableFiles);
    return availableFiles;
  };

  /**
   *
   * @param {string[]} path [...parentDirNames, fileName] must include rootDirHandleName at index 0
   * @param {*} rootDirHandle
   */
  async readFile(path, rootDirHandle) {
    let permission = await hasPermission(rootDirHandle);
    if (!permission) {
      throw new NativeFilePermissionError(
        `Permission required to read ${path}`,
      );
    }
    const { fileHandle } = await this._getFileHandle(path, rootDirHandle);
    try {
      const file = await fileHandle.getFile();
      const textContent = await readFile(file);
      return { file, textContent };
    } catch (error) {
      throw new NativeFSReadError(`file ${path.join('/')} read error`, error);
    }
  }

  /**
   *
   * @param {string[]} path [...parentDirNames, fileName] dont include rootDirHandleName in parentDirNames
   * @param {*} rootDirHandle
   * @param {string} textContent
   */
  async saveFile(path, rootDirHandle, textContent) {
    let permission = await hasPermission(rootDirHandle);
    if (!permission) {
      throw new NativeFilePermissionError(
        `Permission required to save ${path}`,
      );
    }

    let fileHandle;
    let shouldCreateFile = false;
    try {
      ({ fileHandle } = await this._getFileHandle(path, rootDirHandle));
    } catch (error) {
      if (error instanceof NativeFSFileNotFoundError) {
        shouldCreateFile = true;
      } else {
        throw error;
      }
    }

    if (shouldCreateFile) {
      try {
        await createFile(path, rootDirHandle, textContent);
        ({ fileHandle } = await this._getFileHandle(path, rootDirHandle));
      } catch (error) {
        throw new NativeFSWriteError('Unable to create file', error);
      }
    }

    await writeFile(fileHandle, textContent);
  }

  async renameFile(oldPath, newPath, rootDirHandle) {
    if (
      oldPath.slice(0, oldPath.length - 1).join('/') !==
      newPath.slice(0, newPath.length - 1).join('/')
    ) {
      throw new Error('Cannot rename parent directories');
    }

    const oldFile = await this.readFile(oldPath, rootDirHandle);
    await this.saveFile(newPath, rootDirHandle, oldFile.textContent);
    await this.deleteFile(oldPath, rootDirHandle);
  }

  async deleteFile(path, rootDirHandle) {
    const { fileHandle, parentHandles } = await this._getFileHandle(
      path,
      rootDirHandle,
    );
    const parentHandle = getLast(parentHandles);
    await parentHandle.removeEntry(fileHandle.name);
  }

  async listFiles(rootDirHandle) {
    const availableFiles = await this._calculateFilesCache(rootDirHandle);

    return [...availableFiles.values()];
  }
}
/**
 *
 * Note this will always return a list of files and not empty directories
 * @param {Object} dirHandle The directory handle
 * @returns {Array<[dirHandles, fileHandle]>} returns a 2 dimensional array, with each element having [...parentDirHandless, fileHandle].
 *          The parent dir are in order of decreasing order of their distance from file, first parent being the ancestor of all others, and the last parent
 *           being the direct parent of file.
 */
async function recurseDirHandle(
  dirHandle,
  {
    allowedFile = async (fileHandle) => true,
    allowedDir = async (dirHandle) => true,
  } = {},
) {
  let result = [];
  for await (const entry of dirHandle.values()) {
    if (entry.kind === 'file' && allowedFile(entry)) {
      result.push([dirHandle, entry]);
    }
    if (entry.kind === 'directory' && allowedDir(entry)) {
      let children = await recurseDirHandle(entry, { allowedDir, allowedFile });
      // attach the parent first
      children = children.map((r) => [dirHandle, ...r]);
      result = result.concat(children);
    }
  }
  return result.filter((r) => r.length > 0);
}

/**
 * Finds a file given a path also caches it for speedier access.
 * Will throw error if file is not found.
 */
function getFileHandle({
  allowedDir = async (dirHandle) => true,
  allowedFile = async (fileHandle) => true,
}) {
  let dirToChildMap = new WeakMap();
  const getChildHandle = async (childName, dirHandle) => {
    const recalcuteChildren = async () => {
      let children = await asyncIteratorToArray(dirHandle.values());
      children = children.filter((entry) => {
        if (entry.kind === 'directory') {
          return allowedDir(entry);
        }
        if (entry.kind === 'file') {
          return allowedFile(entry);
        }
        throw new Error('Unknown kind of entry: ' + entry.kind);
      });
      dirToChildMap.set(dirHandle, children);
    };

    const findChild = () => {
      const children = dirToChildMap.get(dirHandle);
      if (!children) {
        return undefined;
      }
      const match = children.find((entry) => entry.name === childName);
      return match;
    };

    let match = findChild();

    if (match) {
      return match;
    }
    log('Cache miss');

    await recalcuteChildren();

    return findChild();
  };

  const recurse = async (path, dirHandle, absolutePath, parents) => {
    if (dirHandle.kind !== 'directory') {
      throw new NativeFSReadError(
        `Cannot get Path "${path.join('/')}" as "${
          dirHandle.name
        }" is not a directory`,
      );
    }

    parents.push(dirHandle);

    const [parentName, ...rest] = path;

    if (path.length === 1) {
      return dirHandle
        .getFileHandle(parentName, {
          create: false,
        })
        .catch(handleNotFoundDOMException(absolutePath));
    }

    const handle = await getChildHandle(parentName, dirHandle);

    if (!handle) {
      throw new NativeFSFileNotFoundError(
        `Path "${absolutePath.join('/')}" not found`,
      );
    }

    return recurse(rest, handle, absolutePath, parents);
  };

  return async (path, rootDirHandle) => {
    if (path[0] !== rootDirHandle.name) {
      throw new Error(
        `getFile Error: root parent ${path[0]} must be the rootDirHandle ${rootDirHandle.name}`,
      );
    }

    let parentHandles = [];

    const fileHandle = await recurse(
      path.slice(1),
      rootDirHandle,
      path,
      parentHandles,
    );

    return {
      fileHandle,
      parentHandles,
    };
  };
}

function getLast(array) {
  return array[array.length - 1];
}

function readFile(file) {
  // If the new .text() reader is available, use it.
  if (file.text) {
    return file.text();
  }
  // Otherwise use the traditional file reading technique.
  return _readFileLegacy(file);
}

/**
 *
 * @param {string[]} path [parent, ...rest] array of names relative to dirHandle
 * @param {*} dirHandle
 */
async function createFile(path, rootDirHandle) {
  const recurse = async (path, dirHandle) => {
    const [parentName, ...rest] = path;

    if (path.length === 1) {
      return dirHandle.getFileHandle(parentName, { create: true });
    }

    const newHandle = await dirHandle.getDirectoryHandle(parentName, {
      create: true,
    });

    return recurse(rest, newHandle);
  };

  if (path[0] !== rootDirHandle.name) {
    throw new Error(
      `getFile Error: root parent ${path[0]} must be the rootDirHandle ${rootDirHandle.name}`,
    );
  }
  return recurse(path.slice(1), rootDirHandle);
}

async function writeFile(fileHandle, contents) {
  // Support for Chrome 82 and earlier.
  if (fileHandle.createWriter) {
    // Create a writer (request permission if necessary).
    const writer = await fileHandle.createWriter();
    // Write the full length of the contents
    await writer.write(0, contents);
    // Close the file and write the contents to disk
    await writer.close();
    return;
  }
  // For Chrome 83 and later.
  // Create a FileSystemWritableFileStream to write to.
  const writable = await fileHandle.createWritable();

  // TODO this throws a promise rejection internally, not sure
  // if this the root cause of the crswap issue
  // Write the contents of the file to the stream.
  await writable.write(contents);

  // Close the file and write the contents to disk.
  await writable.close();
}

/**
 * Reads the raw text from a file.
 *
 * @private
 * @param {File} file
 * @return {Promise<string>} A promise that resolves to the parsed string.
 */
function _readFileLegacy(file) {
  return new Promise((resolve) => {
    const reader = new FileReader();
    reader.addEventListener('loadend', (e) => {
      const text = e.srcElement.result;
      resolve(text);
    });
    reader.readAsText(file);
  });
}

export async function pickADirectory(dirHandle) {
  if (dirHandle) {
    let permission = await requestPermission(dirHandle);
    console.log('got permissions');
    if (!permission) {
      throw new NativeFilePermissionError(
        'The permission to edit directory was denied',
      );
    }
  } else {
    try {
      dirHandle = await window.showDirectoryPicker();
      let permission = await requestPermission(dirHandle);
      if (!permission) {
        throw new NativeFilePermissionError(
          'The permission to edit directory was denied',
        );
      }
    } catch (err) {
      console.error(err);
      throw new Error(err.message);
    }
  }

  return dirHandle;
}

export async function hasPermission(dirHandle) {
  const opts = {};
  opts.writable = true;
  // For Chrome 86 and later...
  opts.mode = 'readwrite';
  return (await dirHandle.queryPermission(opts)) === 'granted';
}

export async function requestPermission(dirHandle) {
  const opts = {};
  opts.writable = true;
  // For Chrome 86 and later...
  opts.mode = 'readwrite';
  const perms = await dirHandle.requestPermission(opts);

  return perms === 'granted';
}

async function asyncIteratorToArray(iter) {
  const arr = [];
  for await (const i of iter) {
    arr.push(i);
  }
  return arr;
}

function handleNotFoundDOMException(filePath) {
  return (error) => {
    if (error.name === 'NotFoundError' && error instanceof DOMException) {
      throw new NativeFSFileNotFoundError(
        `Path "${filePath.join('/"')}" not found`,
        error,
      );
    }
    throw error;
  };
}

export class NativeFSReadError extends Error {
  constructor(message, src) {
    super(message);
    if (src) {
      console.log('original error');
      console.error(src);
    }
    this.name = 'NativeFSReadError';
    this.src = src;
  }
}

export class NativeFSWriteError extends Error {
  constructor(message, src) {
    super(message);
    if (src) {
      console.log('original error');
      console.error(src);
    }
    this.name = 'NativeFSWriteError';
    this.src = src;
  }
}

export class NativeFSFileNotFoundError extends NativeFSReadError {
  constructor(msg, src) {
    super(msg, src);
    this.name = 'NativeFSFileNotFoundError';
  }
}

export class NativeFilePermissionError extends NativeFSReadError {
  constructor(message) {
    super(message);
    this.name = 'NativeFilePermissionError';
  }
}
