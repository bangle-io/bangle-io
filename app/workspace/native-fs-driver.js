import { uuid } from '@bangle.dev/core/utils/js-utils';
import { markdownParser, markdownSerializer } from '../markdown/parsers';
import { readFile } from '../misc/index';
const DIR_IGNORE_LIST = ['node_modules', '.git'];

export class FSStorage {
  // TODO check if the new directory is something we already have
  // in our database
  static async createInstance(dirHandle, schema) {
    dirHandle = await pickADirectory(dirHandle);
    if (!(await verifyPermission(dirHandle))) {
      throw new NoPermissionError('Permission denied by user');
    }
    const instance = new FSStorage(dirHandle, schema);
    await instance._updateFilePathHandles();
    return instance;
  }

  static getFilePathKey = (filePathHandles) => {
    return filePathHandles.map((r) => r.name).join('/');
  };

  constructor(dirHandle, schema) {
    this.dirHandle = dirHandle;
    this._schema = schema;
  }

  _getData = async (filePathHandles) => {
    const file = await getLast(filePathHandles).getFile();
    const textContent = await readFile(file);
    const key = FSStorage.getFilePathKey(filePathHandles);
    const doc = markdownParser(textContent);
    return {
      key,
      doc: doc.toJSON(),
      metadata: { lastModified: file.lastModified },
    };
  };

  _updateFilePathHandles = async () => {
    this._filePathHandles = await recurseDirHandle(this.dirHandle, {
      allowedFile: (entry) => entry.name.endsWith('.md'),
      allowedDir: (entry) => !DIR_IGNORE_LIST.includes(entry.name),
    });
    return this._filePathHandles;
  };

  _findPathHandlersByKey = (key) => {
    return this._filePathHandles?.find(
      (paths) => FSStorage.getFilePathKey(paths) === key,
    );
  };

  async iterate(cb) {
    await this._updateFilePathHandles();

    await Promise.all(
      this._filePathHandles.map(async (filePathHandle, i) => {
        const key = FSStorage.getFilePathKey(filePathHandle);
        cb(null, key, i);
      }),
    );
  }

  // TODO support created nested keys
  createNewItemKey(fileName = uuid(6)) {
    // root

    fileName = fileName.endsWith('.md') ? fileName : fileName + '.md';

    return fileName;
  }

  async getItem(key) {
    const match = this._findPathHandlersByKey(key);

    if (match) {
      const { doc, metadata } = await this._getData(match);
      return { doc, metadata };
    }
  }

  async removeItem(key) {
    const match = this._findPathHandlersByKey(key);
    if (!match) {
      throw new Error(`Key ${key} not found`);
    }

    await match[match.length - 2].removeEntry(getLast(match).name);
    await this._updateFilePathHandles();
  }

  async setItem(key, value) {
    const _setItem = async (match) => {
      const handler = match[match.length - 1];
      let data = '';
      // When a new file is created initially the value.doc is null
      if (value.doc) {
        const doc = this._schema.nodeFromJSON(value.doc);
        data = markdownSerializer(doc);
      }
      await writeFile(handler, data);
      return;
    };

    let match = this._filePathHandles.find(
      (paths) => FSStorage.getFilePathKey(paths) === key,
    );

    if (match) {
      await _setItem(match);
      return;
    }

    // TODO currently this only creates rootlevel files
    // if not found attempt to create it
    const name = getLast(key.split('/'));
    await this.dirHandle.getFileHandle(name, { create: true });
    await this._updateFilePathHandles();
    match = this._filePathHandles.find(
      (paths) => FSStorage.getFilePathKey(paths) === key,
    );
    await _setItem(match);
    return;
  }
}

async function writeFile(fileHandle, contents) {
  // Support for Chrome 82 and earlier.
  if (fileHandle.createWriter) {
    // Create a writer (request permission if necessary).
    const writer = await fileHandle.createWriter();
    // Write the full length of the contents
    await writer.write(0, contents);
    // Close the file and write the contents to disk
    await writer.close();
    return;
  }
  // For Chrome 83 and later.
  // Create a FileSystemWritableFileStream to write to.
  const writable = await fileHandle.createWritable();

  // TODO this throws a promise rejection internally, not sure
  // if this the root cause of the crswap issue
  // Write the contents of the file to the stream.
  await writable.write(contents);

  // Close the file and write the contents to disk.
  await writable.close();
}

function getLast(array) {
  return array[array.length - 1];
}

export class NoPermissionError extends Error {
  constructor(errorCode, body) {
    super(body);
    this.errorCode = errorCode;
    this.body = body;
    Error.captureStackTrace(this, NoPermissionError);
    this.name = this.constructor.name;
  }
}

async function pickADirectory(dirHandle) {
  if (dirHandle) {
    let permission = await verifyPermission(dirHandle);
    console.log('got permissions');
    if (!permission) {
      throw new NoPermissionError(
        'The permission to edit directory was denied',
      );
    }
  } else {
    try {
      dirHandle = await window.showDirectoryPicker();
    } catch (err) {
      console.error(err);
      throw new NoPermissionError(err.message);
    }
  }

  return dirHandle;
}

async function verifyPermission(fileHandle, withWrite = true) {
  const opts = {};
  if (withWrite) {
    opts.writable = true;
    // For Chrome 86 and later...
    opts.mode = 'readwrite';
  }
  // Check if we already have permission, if so, return true.
  if ((await fileHandle.queryPermission(opts)) === 'granted') {
    return true;
  }
  // Request permission to the file, if the user grants permission, return true.
  if ((await fileHandle.requestPermission(opts)) === 'granted') {
    return true;
  }
  // The user did nt grant permission, return false.
  return false;
}

/**
 *
 * @param {Object} dirHandle The directory handle
 * @returns {Array<[dirHandles, fileHandle]>} returns a 2 dimensional array, with each element having [...parentDirHandless, fileHandle].
 *          The parent dir are in order of decreasing order of their distance from file, first parent being the ancestor of all others, and the last parent
 *           being the direct parent of file.
 */
async function recurseDirHandle(
  dirHandle,
  {
    allowedFile = async (fileHandle) => true,
    allowedDir = async (dirHandle) => true,
  } = {},
) {
  let result = [];
  for await (const entry of dirHandle.values()) {
    if (entry.kind === 'file' && allowedFile(entry)) {
      result.push([dirHandle, entry]);
    }
    if (entry.kind === 'directory' && allowedDir(entry)) {
      let children = await recurseDirHandle(entry, { allowedDir, allowedFile });
      // attach the parent first
      children = children.map((r) => [dirHandle, ...r]);
      result = result.concat(children);
    }
  }
  return result.filter((r) => r.length > 0);
}

export async function hasPermissions(dirHandle) {
  const opts = {};
  opts.writable = true;
  // For Chrome 86 and later...
  opts.mode = 'readwrite';
  return (await dirHandle.queryPermission(opts)) === 'granted';
}

export async function requestPermission(dirHandle) {
  const opts = {};
  opts.writable = true;
  // For Chrome 86 and later...
  opts.mode = 'readwrite';
  return (await dirHandle.requestPermission(opts)) === 'granted';
}
