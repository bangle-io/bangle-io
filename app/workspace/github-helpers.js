import Github from 'github-api';
import * as _idb from 'idb-keyval';
import './nativefs-helpers';
import { resolvePath } from './path-helpers';

const urlParams = new URLSearchParams(window.location.search);

const z = new Github({
  token: urlParams.get('token'),
});

let getRepoName = (wsPath) => {
  let { wsName, filePath } = resolvePath(wsPath);
  if (wsName.startsWith('github--')) {
    return wsName.split('github--').join('');
  }
  return wsName;
};
export class GithubFS {
  // file
  static async getFile(wsPath, github) {
    let { wsName, filePath } = resolvePath(wsPath);

    const repo = await z.getRepo('kepta', getRepoName(wsPath));

    console.log(filePath);
    const { data } = await repo.getContents('master', filePath, true);

    return data;
  }

  static async deleteFile(wsPath) {}

  static async renameFile(wsPath, newWsPath) {}

  static async createFile(wsPath, payload) {}

  static async updateFile(wsPath, payload) {}

  static async listFiles(wsName, github) {
    const repo = await z.getRepo('kepta', getRepoName(wsName + ':' + 'sd'));

    const branch = await repo.getBranch('master');
    const sha = branch.data.commit.sha;
    const { tree } = (await repo.getTree(sha)).data;
    const result = tree.map((r) => wsName + ':' + r.path);
    return result;
  }
}
