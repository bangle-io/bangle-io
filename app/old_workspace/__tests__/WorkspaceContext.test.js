import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { IndexDbWorkspace } from 'bangle-io/app/workspace/workspace';
import { IndexDbWorkspaceFile } from 'bangle-io/app/workspace/workspace-file';

import localforage from 'localforage';
import {
  workspaceActions,
  WorkspaceContext,
  WorkspaceContextProvider,
} from '../WorkspaceContext';
import { sleep } from '@bangle.dev/core/utils/js-utils';
import { INDEXDB_TYPE } from 'bangle-io/app/workspace/type-helpers';
import { SpecRegistry } from '@bangle.dev/core/spec-registry';

jest.mock('localforage', () => {
  const instance = {
    iterate: jest.fn(async () => {}),
    setItem: jest.fn(async () => {}),
    removeItem: jest.fn(async () => {}),
    getItem: jest.fn(async () => {}),
  };
  const workspacesInstance = {
    iterate: jest.fn(async () => {}),
    setItem: jest.fn(async () => {}),
    removeItem: jest.fn(async () => {}),
    getItem: jest.fn(async () => {}),
  };
  return {
    config: jest.fn(),
    createInstance: jest.fn(({ name } = {}) => {
      if (name === 'workspaces/1') {
        return workspacesInstance;
      }
      return instance;
    }),
  };
});
const DateNowBackup = jest.fn();

describe('index db workspace', () => {
  let dbInstance;
  const schema = new SpecRegistry().schema;
  beforeEach(async () => {
    Date.now = jest.fn(() => 1);
    dbInstance = localforage.createInstance();
    dbInstance.getItem = jest.fn(async (docName) => ({ docName, doc: null }));
    dbInstance.iterate.mockImplementationOnce(async (cb) => {
      Array.from({ length: 5 }, (_, k) =>
        cb(
          {
            docName: k + '',
            doc: null,
          },
          k + '',
          k,
        ),
      );
    });
  });
  afterEach(() => {
    Date.now = DateNowBackup;
  });
  test('Creates workspace correctly', async () => {
    await IndexDbWorkspace.createWorkspace('test_db', schema, INDEXDB_TYPE);
  });

  test('Adds files correctly', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace(
      'test_db',
      schema,
      INDEXDB_TYPE,
    );

    expect(workspace.files).toMatchSnapshot();
  });

  test('getFile', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace(
      'test_db',
      schema,
      INDEXDB_TYPE,
    );

    expect(workspace.getFile('1')).toMatchInlineSnapshot(`
      Object {
        "doc": null,
        "docName": "1",
        "metadata": Object {
          "lastModified": 1,
        },
      }
    `);
  });

  test('hasFile', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace(
      'test_db',
      schema,
      INDEXDB_TYPE,
    );

    expect(workspace.hasFile('2')).toBe(true);
  });

  test('link file', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace(
      'test_db',
      schema,
      INDEXDB_TYPE,
    );

    const newFile = new IndexDbWorkspaceFile(
      'test_doc',
      null,
      {},
      { dbInstance: dbInstance, schema: schema },
    );

    let newWorkspace = workspace.linkFile(newFile);
    expect(workspace.hasFile('test_doc')).toBe(false);
    expect(newWorkspace.hasFile('test_doc')).toBe(true);
  });

  test('unlink file', async () => {
    const workspace = await IndexDbWorkspace.createWorkspace(
      'test_db',
      schema,
      INDEXDB_TYPE,
    );

    let newWorkspace = workspace.unlinkFile(workspace.getFile('2'));

    expect(workspace.hasFile('2')).toBe(true);
    expect(newWorkspace.hasFile('2')).toBe(false);
  });
});

describe('workspaceContext actions', () => {
  let dbInstance;
  let workspaceInstance;
  const customRender = (child, { ...renderOptions } = {}) => {
    return render(
      <WorkspaceContextProvider>{child}</WorkspaceContextProvider>,
      renderOptions,
    );
  };

  beforeEach(async () => {});
  afterEach(() => {
    Date.now = DateNowBackup;
  });

  beforeEach(async () => {
    Date.now = jest.fn(() => 1);
    dbInstance = localforage.createInstance();
    workspaceInstance = localforage.createInstance({ name: 'workspaces/1' });
    dbInstance.getItem = jest.fn(async (docName) => ({ docName, doc: null }));
    dbInstance.iterate.mockImplementation(async (cb) => {
      Array.from({ length: 5 }, (_, k) =>
        cb(
          {
            docName: k + '',
            doc: null,
          },
          k + '',
          k,
        ),
      );
    });
  });

  test('onMountWorkspaceLoad no existing workspaceInfo', async () => {
    workspaceInstance.getItem.mockImplementation(() => undefined);
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace }) =>
          workspace && (
            <span data-testid="result">
              Result: {workspace.files.map((f) => f.docName)}
            </span>
          )
        }
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result').textContent).toMatchInlineSnapshot(
      `"Result: 01234"`,
    );
    expect(dbInstance.getItem).toBeCalledTimes(5);
  });

  test('onMountWorkspaceLoad when there is existing workspaceInfo', async () => {
    const name = 'test1';
    workspaceInstance.getItem.mockImplementation(async () => ({
      indexdb_abc1: {
        metadata: {},
        uid: 'indexdb_abc1',
        name,
      },
    }));
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace }) =>
          workspace && (
            <span data-testid="result">
              Result: {workspace.name} {workspace.files.map((f) => f.docName)}
            </span>
          )
        }
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result').textContent).toEqual(
      `Result: ${name} 01234`,
    );
    expect(dbInstance.getItem).toBeCalledTimes(5);
  });

  test('takeWorkspaceBackup', async () => {
    const fixture = await import('./backup-fixture.json');
    let _updateContext;
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _updateContext = updateContext;
          return (
            <>
              <span data-testid="result">
                Result: {workspace.name} {workspace.files.map((f) => f.docName)}
              </span>
              <span data-testid="result2">
                {workspace.files.map((f, key) => (
                  <span key={key}>{f.title}</span>
                ))}
              </span>
            </>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    dbInstance.getItem.mockImplementation(() => undefined);

    await _updateContext(
      workspaceActions.newWorkspaceFromBackup(fixture, 'indexdb'),
    );

    await sleep(500);

    expect(screen.getByTestId('result').textContent).toEqual(
      `Result: backup-fixture fixture_doc_2fixture_doc_1`,
    );
    expect(screen.getByTestId('result2')).toMatchInlineSnapshot(`
      <span
        data-testid="result2"
      >
        <span>
          Hello
        </span>
        <span>
          I am second document
        </span>
      </span>
    `);
  });

  test('has updateContext', async () => {
    customRender(
      <WorkspaceContext.Consumer>
        {({ updateContext } = {}) =>
          updateContext && (
            <span data-testid="result">{Boolean(updateContext)}</span>
          )
        }
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result')).toBeTruthy();
  });

  test('has one openedDocuments initially', async () => {
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments } = {}) => {
          if (!workspace) {
            return;
          }
          expect(openedDocuments.every((r) => r.docName)).toBe(true);
          expect(openedDocuments.every((r) => r.key)).toBe(true);
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    expect(screen.getByTestId('result')).toMatchInlineSnapshot(`
      <span
        data-testid="result"
      >
        0
      </span>
    `);
  });

  test('openWorkspaceFile', async () => {
    let _updateContext;
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _updateContext = updateContext;
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    await _updateContext(workspaceActions.openWorkspaceFile('2'));

    expect(screen.getByTestId('result')).toMatchInlineSnapshot(`
      <span
        data-testid="result"
      >
        2
        0
      </span>
    `);

    // opening again should work
    await _updateContext(workspaceActions.openWorkspaceFile('2'));

    expect(screen.getByTestId('result').textContent).toMatchInlineSnapshot(
      `"22"`,
    );
  });

  test('openBlankWorkspaceFile', async () => {
    let _updateContext;
    let _openedDocuments;
    let _workspace;

    dbInstance.iterate.mockImplementation(async (cb) => {
      Array.from({ length: 2 }, (_, k) =>
        cb(
          {
            docName: k + '',
            doc: null,
          },
          k + '',
          k,
        ),
      );
    });

    dbInstance.getItem = jest.fn(async (docName) => {
      // we only give out 2 files in mock iterate
      if (['0', '1'].includes(docName)) {
        return {
          docName,
          doc: null,
        };
      }
    });

    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _workspace = workspace;
          _updateContext = updateContext;
          _openedDocuments = openedDocuments;
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    await _updateContext(workspaceActions.openBlankWorkspaceFile());

    expect(_openedDocuments.length).toBe(2);
    // its a random string
    expect(typeof _openedDocuments[0].docName === 'string').toBe(true);

    expect(_workspace.hasFile(_openedDocuments[0].docName)).toBe(true);
    expect(dbInstance.setItem).toHaveBeenNthCalledWith(
      1,
      _openedDocuments[0].docName,
      expect.any(Object),
    );
  });

  test('deleteWorkspaceFile', async () => {
    let _updateContext;
    let _openedDocuments;
    let _workspace;
    customRender(
      <WorkspaceContext.Consumer>
        {({ workspace, openedDocuments, updateContext } = {}) => {
          if (!workspace) {
            return;
          }
          _workspace = workspace;
          _updateContext = updateContext;
          _openedDocuments = openedDocuments;
          return (
            <span data-testid="result">
              {openedDocuments.map((r) => r.docName)}
            </span>
          );
        }}
      </WorkspaceContext.Consumer>,
    );

    await waitFor(() => screen.getByTestId('result'));

    await _updateContext(workspaceActions.openWorkspaceFile('2'));
    expect(_openedDocuments[0].docName).toBe('2');

    await _updateContext(workspaceActions.deleteWorkspaceFile('2'));

    expect(_openedDocuments.length).toBe(1);
    // its a random string
    expect(_openedDocuments[0].docName).toBe('0');

    expect(_workspace.hasFile(_openedDocuments[0].docName)).toBe(true);
    expect(dbInstance.removeItem).toBeCalledTimes(1);
    expect(dbInstance.removeItem).toHaveBeenNthCalledWith(1, '2');
  });
});

test.todo('Be able to delete a pending workspace');

describe('restore native-fs', () => {
  test.todo('take a backup');
  test.todo('Be able to restore to a folder with already existing files');
  test.todo(
    'Be able to restore to a folder with already existing files and has a name conflict',
  );

  test.todo('Have multiple workspaces pointing to same directory');
  test.todo(
    'Have multiple workspaces  of which one directory is a child or the other workspace',
  );

  test.todo('Handle the case when the directory backing workspace is deleted');
});
