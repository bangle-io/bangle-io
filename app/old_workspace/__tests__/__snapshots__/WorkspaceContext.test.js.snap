// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index db workspace Adds files correctly 1`] = `
Array [
  Object {
    "doc": null,
    "docName": "0",
    "metadata": Object {
      "lastModified": 1,
    },
  },
  Object {
    "doc": null,
    "docName": "1",
    "metadata": Object {
      "lastModified": 1,
    },
  },
  Object {
    "doc": null,
    "docName": "2",
    "metadata": Object {
      "lastModified": 1,
    },
  },
  Object {
    "doc": null,
    "docName": "3",
    "metadata": Object {
      "lastModified": 1,
    },
  },
  Object {
    "doc": null,
    "docName": "4",
    "metadata": Object {
      "lastModified": 1,
    },
  },
]
`;
